File: utils/visualization.R
library(ggplot2)
library(ggpubr)
library(dplyr)
library(tidyr)
library(viridis)
library(reshape2)

# Function to visualize confusion matrix
plot_confusion_matrix <- function(conf_matrix, class_labels) {
  conf_matrix_melt <- melt(conf_matrix)
  colnames(conf_matrix_melt) <- c("Predicted", "Actual", "Count")
  
  ggplot(conf_matrix_melt, aes(x = Predicted, y = Actual, fill = Count)) +
    geom_tile(color = "white") +
    scale_fill_viridis(option = "C") +
    geom_text(aes(label = sprintf("%d", Count)), vjust = 1, color = "white") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(
      title = "Confusion Matrix",
      x = "Predicted Class",
      y = "Actual Class",
      fill = "Count"
    )
}

# Function to plot ROC curve
plot_roc_curve <- function(roc_data) {
  ggplot(roc_data, aes(x = FPR, y = TPR, color = Model)) +
    geom_line(size = 1) +
    scale_color_viridis(discrete = TRUE) +
    theme_minimal() +
    labs(
      title = "ROC Curve",
      x = "False Positive Rate",
      y = "True Positive Rate",
      color = "Model"
    ) +
    geom_abline(linetype = "dashed", color = "gray")
}

# Function to plot precision-recall curve
plot_pr_curve <- function(pr_data) {
  ggplot(pr_data, aes(x = Recall, y = Precision, color = Model)) +
    geom_line(size = 1) +
    scale_color_viridis(discrete = TRUE) +
    theme_minimal() +
    labs(
      title = "Precision-Recall Curve",
      x = "Recall",
      y = "Precision",
      color = "Model"
    )
}

# Function to plot accuracy and loss over epochs
plot_training_metrics <- function(metrics_data) {
  metrics_data_long <- gather(metrics_data, key = "Metric", value = "Value", -Epoch)
  
  ggplot(metrics_data_long, aes(x = Epoch, y = Value, color = Metric)) +
    geom_line(size = 1) +
    scale_color_viridis(discrete = TRUE) +
    theme_minimal() +
    labs(
      title = "Model Training Metrics",
      x = "Epoch",
      y = "Value",
      color = "Metric"
    )
}

# Function to visualize sentiment distribution
plot_sentiment_distribution <- function(sentiment_data) {
  ggplot(sentiment_data, aes(x = Sentiment, fill = Sentiment)) +
    geom_bar() +
    scale_fill_viridis(discrete = TRUE) +
    theme_minimal() +
    labs(
      title = "Sentiment Distribution",
      x = "Sentiment",
      y = "Count",
      fill = "Sentiment"
    )
}

# Function to visualize word cloud from text data
plot_wordcloud <- function(word_freq_data) {
  wordcloud2::wordcloud2(word_freq_data, color = "random-light", backgroundColor = "white")
}

# Function call
# plot_confusion_matrix(conf_matrix, class_labels)
# plot_roc_curve(roc_data)
# plot_pr_curve(pr_data)
# plot_training_metrics(metrics_data)
# plot_sentiment_distribution(sentiment_data)
# plot_wordcloud(word_freq_data)

File: utils/visualization.py
import matplotlib.pyplot as plt
import seaborn as sns
from wordcloud import WordCloud

def plot_training_curves(history, metrics=['loss', 'accuracy']):
    """
    Plots the training and validation curves for specified metrics.
    
    Args:
        history: Training history object containing metric data.
        metrics: List of metrics to plot (e.g., 'loss', 'accuracy').
    """
    epochs = range(1, len(history.history[metrics[0]]) + 1)
    
    for metric in metrics:
        train_values = history.history[metric]
        val_values = history.history[f'val_{metric}']
        
        plt.figure(figsize=(10, 6))
        plt.plot(epochs, train_values, 'bo-', label=f'Training {metric}')
        plt.plot(epochs, val_values, 'r*-', label=f'Validation {metric}')
        plt.title(f'Training and Validation {metric.capitalize()}')
        plt.xlabel('Epochs')
        plt.ylabel(metric.capitalize())
        plt.legend()
        plt.grid(True)
        plt.show()

def generate_word_cloud(text, max_words=100, colormap='viridis'):
    """
    Generates and displays a word cloud from a given text.
    
    Args:
        text: Input text data for generating the word cloud.
        max_words: Maximum number of words to include in the word cloud.
        colormap: Color map for the word cloud visualization.
    """
    wordcloud = WordCloud(width=800, height=400, background_color='white', max_words=max_words, colormap=colormap).generate(text)
    
    plt.figure(figsize=(10, 6))
    plt.imshow(wordcloud, interpolation='bilinear')
    plt.axis('off')
    plt.show()

def plot_confusion_matrix(confusion_matrix, class_names):
    """
    Plots a confusion matrix heatmap.
    
    Args:
        confusion_matrix: Confusion matrix to visualize.
        class_names: List of class names to annotate the matrix.
    """
    plt.figure(figsize=(8, 6))
    sns.heatmap(confusion_matrix, annot=True, fmt='d', cmap='Blues', xticklabels=class_names, yticklabels=class_names)
    plt.xlabel('Predicted')
    plt.ylabel('Actual')
    plt.title('Confusion Matrix')
    plt.show()